### ISSUE TITLE
Missing lint for mismatched type on @override?

### ISSUE THREAD

**Issue by eseidel** [dart-lang/linter#2563]
**Describe the rule you'd like to see implemented**

I would expect the analyzer to warn when you @override a method and have a mismatched return type, including nullable? Maybe this lint already exists?

**Examples**

Overriding with the wrong return type in `Sub` should warn:
```dart
class Super {
  int? method() => 1;
}

class Sub extends Super {
  @override
  int method() => 2;
}

void main() {
  print(Super().method());
  print(Sub().method());
}
```

What actually motivated this was working through:
https://medium.com/flutter/learning-flutters-new-navigation-and-routing-system-7c9068155ade

And having copied:
```dart
  @override
  RouteInformation restoreRouteInformation(BookRoutePath path) {
    if (path.isUnknown) {
      return RouteInformation(location: '/404');
    }
    if (path.isHomePage) {
      return RouteInformation(location: '/');
    }
    if (path.isDetailsPage) {
      return RouteInformation(location: '/book/${path.id}');
    }
    return null;
  }
```
into my project.  But that has the wrong return type!  (I originally removed the null-return path as part of attempting to make it null-safe, not realizing that I needed to change the return type instead.)  Which then resulted in me being confused as to how to use the API correctly.

This isn't a big deal. :)  But I was surprised that even once I turned on package:pedantic, I didn't see any warnings about this mis-matched type on the return.

**Comment by a14n** [comment]
I suspect there would be a lot of false positives. It's fine to have override with sharper return type or widen parameter types.

```dart
abstract class Animal {
  Animal get child;
  bool like(Cat cat);
}
abstract class Cat extends Animal {
  @override
  Cat get child;
  @override
  bool like(Animal otherAnimals);
}
```

Thus if you have an instance of `Cat` you have better information that its `child` is not only an `Animal` but also a `Cat`.

I haven't check on Flutter codebase but I guess this happens often.

**Comment by eseidelGoogle** [comment]
Makes sense.  Closing.
